File: java_projects/glue_code/commons-codec/src/main/java/org/apache/commons/codec/net/URLCodec.java
package org.apache.commons.codec.net;

import org.graalvm.polyglot.Value;
import org.graalvm.polyglot.PolyglotException;
import org.apache.commons.codec.ContextInitializer;
import org.apache.commons.codec.ExceptionHandler;
import org.apache.commons.codec.IntegrationUtils;
import java.io.ByteArrayOutputStream;
import java.io.UnsupportedEncodingException;
import java.util.BitSet;
import org.apache.commons.codec.BinaryDecoder;
import org.apache.commons.codec.DecoderException;
import org.apache.commons.codec.BinaryEncoder;
import org.apache.commons.codec.EncoderException;
import org.apache.commons.codec.CharEncoding;
import org.apache.commons.codec.StringDecoder;
import org.apache.commons.codec.StringEncoder;
import org.apache.commons.codec.binary.StringUtils;
public class URLCodec implements BinaryEncoder, BinaryDecoder, StringEncoder, StringDecoder {
    protected static final byte ESCAPE_CHAR = '%';
    private static final BitSet WWW_FORM_URL_SAFE = new BitSet(256);
    private static Value clz = ContextInitializer.getPythonClass("/net/URLCodec.py", "URLCodec");
    private Value obj;
    public URLCodec(Value obj) {
        this.obj = obj;
    }
    public Value getPythonObject() {
        return obj;
    }
    public String getEncoding() {
// 
// return this.charset;
// 


// TODO: Check the type mapping below!
return obj.invokeMember("getEncoding").as(String.class);
}
    public Object decode(final Object obj) throws DecoderException {
try {
// 
// return decode3(obj);
// 


// TODO: Check the type mapping below!
return obj.invokeMember("decode", obj).as(Object.class);
} catch (PolyglotException e) {
    // TODO: Handle DecoderException
    throw (DecoderException) ExceptionHandler.handle(e, "URLCodec.decode");
}
}
    public Object encode(final Object obj) throws EncoderException {
try {
// 
// return encode3(obj);
// 


// TODO: Check the type mapping below!
return obj.invokeMember("encode", obj).as(Object.class);
} catch (PolyglotException e) {
    // TODO: Handle EncoderException
    throw (EncoderException) ExceptionHandler.handle(e, "URLCodec.encode");
}
}
    public String decode(final String str) throws DecoderException {
try {
// 
// return decode2(str);
// 


// TODO: Check the type mapping below!
return obj.invokeMember("decode", str).as(String.class);
} catch (PolyglotException e) {
    // TODO: Handle DecoderException
    throw (DecoderException) ExceptionHandler.handle(e, "URLCodec.decode");
}
}
    public String encode(final String str) throws EncoderException {
try {
// 
// return encode2(str);
// 


// TODO: Check the type mapping below!
return obj.invokeMember("encode", str).as(String.class);
} catch (PolyglotException e) {
    // TODO: Handle EncoderException
    throw (EncoderException) ExceptionHandler.handle(e, "URLCodec.encode");
}
}
    public byte[] decode(final byte[] bytes) throws DecoderException {
try {
// 
// return decode0(bytes);
// 


// TODO: Check the type mapping below!
return obj.invokeMember("decode", bytes).as(byte[].class);
} catch (PolyglotException e) {
    // TODO: Handle DecoderException
    throw (DecoderException) ExceptionHandler.handle(e, "URLCodec.decode");
}
}
    public byte[] encode(final byte[] bytes) {
// 
// return encode0(bytes);
// 


// TODO: Check the type mapping below!
return obj.invokeMember("encode", bytes).as(byte[].class);
}
    public String getDefaultCharset() {
// 
// return this.charset;
// 


// TODO: Check the type mapping below!
return obj.invokeMember("getDefaultCharset").as(String.class);
}
    public Object decode3(final Object obj) throws DecoderException {
try {
// 
// if (obj == null) {
// return null;
// }
// if (obj instanceof byte[]) {
// return decode0((byte[]) obj);
// }
// if (obj instanceof String) {
// return decode2((String) obj);
// }
// throw new DecoderException(
// "Objects of type " + obj.getClass().getName() + " cannot be URL decoded", null);
// 


// TODO: Check the type mapping below!
return obj.invokeMember("decode3", obj).as(Object.class);
} catch (PolyglotException e) {
    // TODO: Handle DecoderException
    throw (DecoderException) ExceptionHandler.handle(e, "URLCodec.decode3");
}
}
    public Object encode3(final Object obj) throws EncoderException {
try {
// 
// if (obj == null) {
// return null;
// }
// if (obj instanceof byte[]) {
// return encode0((byte[]) obj);
// }
// if (obj instanceof String) {
// return encode2((String) obj);
// }
// throw new EncoderException(
// "Objects of type " + obj.getClass().getName() + " cannot be URL encoded", null);
// 


// TODO: Check the type mapping below!
return obj.invokeMember("encode3", obj).as(Object.class);
} catch (PolyglotException e) {
    // TODO: Handle EncoderException
    throw (EncoderException) ExceptionHandler.handle(e, "URLCodec.encode3");
}
}
    public String decode2(final String str) throws DecoderException {
try {
// 
// if (str == null) {
// return null;
// }
// try {
// return decode1(str, getDefaultCharset());
// } catch (final UnsupportedEncodingException e) {
// throw new DecoderException(e.getMessage(), e);
// }
// 


// TODO: Check the type mapping below!
return obj.invokeMember("decode2", str).as(String.class);
} catch (PolyglotException e) {
    // TODO: Handle DecoderException
    throw (DecoderException) ExceptionHandler.handle(e, "URLCodec.decode2");
}
}
    public String decode1(final String str, final String charsetName)
            throws DecoderException, UnsupportedEncodingException {
try {
// 
// if (str == null) {
// return null;
// }
// return new String(decode0(StringUtils.getBytesUsAscii(str)), charsetName);
// 


// TODO: Check the type mapping below!
return obj.invokeMember("decode1", str, charsetName).as(String.class);
} catch (PolyglotException e) {
    // TODO: Handle DecoderException, UnsupportedEncodingException
    throw (DecoderException, UnsupportedEncodingException) ExceptionHandler.handle(e, "URLCodec.decode1");
}
}
    public String decode(final String str, final String charsetName)
            throws DecoderException, UnsupportedEncodingException {
try {
// 
// return decode1(str, charsetName);
// 


// TODO: Check the type mapping below!
return obj.invokeMember("decode", str, charsetName).as(String.class);
} catch (PolyglotException e) {
    // TODO: Handle DecoderException, UnsupportedEncodingException
    throw (DecoderException, UnsupportedEncodingException) ExceptionHandler.handle(e, "URLCodec.decode");
}
}
    public String encode2(final String str) throws EncoderException {
try {
// 
// if (str == null) {
// return null;
// }
// try {
// return encode1(str, getDefaultCharset());
// } catch (final UnsupportedEncodingException e) {
// throw new EncoderException(e.getMessage(), e);
// }
// 


// TODO: Check the type mapping below!
return obj.invokeMember("encode2", str).as(String.class);
} catch (PolyglotException e) {
    // TODO: Handle EncoderException
    throw (EncoderException) ExceptionHandler.handle(e, "URLCodec.encode2");
}
}
    public String encode1(final String str, final String charsetName)
            throws UnsupportedEncodingException {
try {
// 
// if (str == null) {
// return null;
// }
// return StringUtils.newStringUsAscii(encode0(str.getBytes(charsetName)));
// 


// TODO: Check the type mapping below!
return obj.invokeMember("encode1", str, charsetName).as(String.class);
} catch (PolyglotException e) {
    // TODO: Handle UnsupportedEncodingException
    throw (UnsupportedEncodingException) ExceptionHandler.handle(e, "URLCodec.encode1");
}
}
    public String encode(final String str, final String charsetName)
            throws UnsupportedEncodingException {
try {
// 
// return encode1(str, charsetName);
// 


// TODO: Check the type mapping below!
return obj.invokeMember("encode", str, charsetName).as(String.class);
} catch (PolyglotException e) {
    // TODO: Handle UnsupportedEncodingException
    throw (UnsupportedEncodingException) ExceptionHandler.handle(e, "URLCodec.encode");
}
}
    public byte[] decode0(final byte[] bytes) throws DecoderException {
try {
// 
// return decodeUrl(bytes);
// 


// TODO: Check the type mapping below!
return obj.invokeMember("decode0", bytes).as(byte[].class);
} catch (PolyglotException e) {
    // TODO: Handle DecoderException
    throw (DecoderException) ExceptionHandler.handle(e, "URLCodec.decode0");
}
}
    public byte[] encode0(final byte[] bytes) {
// 
// return encodeUrl(WWW_FORM_URL_SAFE, bytes);
// 


// TODO: Check the type mapping below!
return obj.invokeMember("encode0", bytes).as(byte[].class);
}
    public static final byte[] decodeUrl(final byte[] bytes) throws DecoderException {
try {
// 
// if (bytes == null) {
// return null;
// }
// final ByteArrayOutputStream buffer = new ByteArrayOutputStream();
// for (int i = 0; i < bytes.length; i++) {
// final int b = bytes[i];
// if (b == '+') {
// buffer.write(' ');
// } else if (b == ESCAPE_CHAR) {
// try {
// final int u = Utils.digit16(bytes[++i]);
// final int l = Utils.digit16(bytes[++i]);
// buffer.write((char) ((u << 4) + l));
// } catch (final ArrayIndexOutOfBoundsException e) {
// throw new DecoderException("Invalid URL encoding: ", e);
// }
// } else {
// buffer.write(b);
// }
// }
// return buffer.toByteArray();
// 


// TODO: Check the type mapping below!
return clz.invokeMember("decodeUrl", bytes).as(byte[].class);
} catch (PolyglotException e) {
    // TODO: Handle DecoderException
    throw (DecoderException) ExceptionHandler.handle(e, "URLCodec.decodeUrl");
}
}
    public static final byte[] encodeUrl(BitSet urlsafe, final byte[] bytes) {
// 
// if (bytes == null) {
// return null;
// }
// if (urlsafe == null) {
// urlsafe = WWW_FORM_URL_SAFE;
// }
// 
// final ByteArrayOutputStream buffer = new ByteArrayOutputStream();
// for (final byte c : bytes) {
// int b = c;
// if (b < 0) {
// b = 256 + b;
// }
// if (urlsafe.get(b)) {
// if (b == ' ') {
// b = '+';
// }
// buffer.write(b);
// } else {
// buffer.write(ESCAPE_CHAR);
// final char hex1 = Utils.hexDigit(b >> 4);
// final char hex2 = Utils.hexDigit(b);
// buffer.write(hex1);
// buffer.write(hex2);
// }
// }
// return buffer.toByteArray();
// 


// TODO: Check the type mapping below!
return clz.invokeMember("encodeUrl", urlsafe, bytes).as(byte[].class);
}
    public static URLCodec URLCodec1() {
// 
// return new URLCodec(CharEncoding.UTF_8);
// 


// TODO: Check the type mapping below!
return clz.invokeMember("URLCodec1").as(URLCodec.class);
}
    public URLCodec(final String charset) {
// 
// this.charset = charset;
// 

this.obj = clz.invokeMember("__init__", charset);
}
}

File: data/recomposed_projects/llama-3-70b-instruct/signature/commons-codec/src/main/org/apache/commons/codec/net/URLCodec.py
from __future__ import annotations
import copy
import re
from io import BytesIO
import io
import typing
from typing import *
from src.main.org.apache.commons.codec.BinaryDecoder import *
from src.main.org.apache.commons.codec.BinaryEncoder import *
from src.main.org.apache.commons.codec.CharEncoding import *
from src.main.org.apache.commons.codec.DecoderException import *
from src.main.org.apache.commons.codec.EncoderException import *
from src.main.org.apache.commons.codec.StringDecoder import *
from src.main.org.apache.commons.codec.StringEncoder import *
from src.main.org.apache.commons.codec.binary.StringUtils import *
from src.main.org.apache.commons.codec.net.Utils import *


class URLCodec(BinaryDecoder, BinaryEncoder, StringDecoder, StringEncoder):

    _WWW_FORM_URL: typing.List[bool] = None

    _ESCAPE_CHAR: int = 37
    _charset: str = ""

    __WWW_FORM_URL_SAFE: typing.List[bool] = [False] * 256

    @staticmethod
    def run_static_init():
        for i in range(ord("a"), ord("z") + 1):
            URLCodec.__WWW_FORM_URL_SAFE[i] = True
        for i in range(ord("A"), ord("Z") + 1):
            URLCodec.__WWW_FORM_URL_SAFE[i] = True
        for i in range(ord("0"), ord("9") + 1):
            URLCodec.__WWW_FORM_URL_SAFE[i] = True
        URLCodec.__WWW_FORM_URL_SAFE[ord("-")] = True
        URLCodec.__WWW_FORM_URL_SAFE[ord("_")] = True
        URLCodec.__WWW_FORM_URL_SAFE[ord(".")] = True
        URLCodec.__WWW_FORM_URL_SAFE[ord("*")] = True
        URLCodec.__WWW_FORM_URL_SAFE[ord(" ")] = True

        URLCodec._WWW_FORM_URL = URLCodec.__WWW_FORM_URL_SAFE.copy()

    def getEncoding(self) -> str:
        return self._charset

    def getDefaultCharset(self) -> str:
        return self._charset

    def decode3(self, obj: typing.Any) -> typing.Any:
        if obj is None:
            return None
        if isinstance(obj, bytes):
            return self.decode0(obj)
        if isinstance(obj, str):
            return self.decode2(obj)
        raise DecoderException(
            f"Objects of type {obj.__class__.__name__} cannot be URL decoded", None
        )

    def encode3(self, obj: typing.Any) -> typing.Any:
        if obj is None:
            return None
        if isinstance(obj, bytes):
            return self.encode0(obj)
        if isinstance(obj, str):
            return self.encode2(obj)
        raise EncoderException(
            f"Objects of type {obj.__class__.__name__} cannot be URL encoded", None
        )

    def decode2(self, str: str) -> str:
        if str is None:
            return None
        try:
            return self.decode1(str, self.getDefaultCharset())
        except ValueError as e:
            raise DecoderException(e.getMessage(), e)

    def decode1(self, str: str, charsetName: str) -> str:
        if str is None:
            return None
        return str(self.decode0(StringUtils.getBytesUsAscii(str)), charsetName)

    def encode2(self, str: str) -> str:
        if str is None:
            return None
        try:
            return self.encode1(str, self.getDefaultCharset())
        except ValueError as e:
            raise EncoderException(e.getMessage(), e)

    def encode1(self, str: str, charsetName: str) -> str:
        if str is None:
            return None
        return StringUtils.newStringUsAscii(self.encode0(str.encode(charsetName)))

    def decode0(self, bytes: typing.List[int]) -> typing.List[int]:
        return URLCodec.decodeUrl(bytes)

    def encode0(self, bytes: typing.List[int]) -> typing.List[int]:
        return URLCodec.encodeUrl(URLCodec.__WWW_FORM_URL_SAFE, bytes)

    @staticmethod
    def decodeUrl(bytes: typing.List[int]) -> typing.List[int]:
        if bytes is None:
            return None
        buffer = io.BytesIO()
        for i in range(len(bytes)):
            b = bytes[i]
            if b == ord("+"):
                buffer.write(ord(" "))
            elif b == URLCodec._ESCAPE_CHAR:
                try:
                    u = Utils.digit16(bytes[i + 1])
                    l = Utils.digit16(bytes[i + 2])
                    buffer.write((u << 4) + l)
                except IndexError as e:
                    raise DecoderException("Invalid URL encoding: ", e)
            else:
                buffer.write(b)
        return buffer.getvalue()

    @staticmethod
    def encodeUrl(
        urlsafe: typing.List[bool], bytes: typing.List[int]
    ) -> typing.List[int]:
        if bytes is None:
            return None
        if urlsafe is None:
            urlsafe = URLCodec.__WWW_FORM_URL_SAFE
        buffer = io.BytesIO()
        for c in bytes:
            b = c
            if b < 0:
                b = 256 + b
            if urlsafe[b]:
                if b == ord(" "):
                    b = ord("+")
                buffer.write(bytes([b]))
            else:
                buffer.write(bytes([URLCodec._ESCAPE_CHAR]))
                hex1 = Utils.hexDigit(b >> 4)
                hex2 = Utils.hexDigit(b)
                buffer.write(bytes([ord(hex1)]))
                buffer.write(bytes([ord(hex2)]))
        return buffer.getvalue()

    @staticmethod
    def URLCodec1() -> URLCodec:
        return URLCodec(CharEncoding.UTF_8)

    def __init__(self, charset: str) -> None:
        self._charset = charset


URLCodec.run_static_init()
